torch/_inductor/codegen/triton.py

```diff
@@ -1210,20 +1207,34 @@ def codegen_block_ptr_store_line(self, name, indexing, block_ptr, value, other="
         value = f"{value}.to({triton_store_type(V.graph.get_dtype(name))})"
         return f"tl.store({block_ptr}, {value}{other})"
 
-    def load_mask(self, var):
-        mask = ""
-        mask_vars = set(var.mask_vars)
-        if self._load_mask:
-            mask_vars.add(self._load_mask)
+    def check_bounds(
+        self,
+        expr: sympy.Expr,
+        size: sympy.Expr,
+        lower: bool,
+        upper: bool,
+    ):
+        if not (lower or upper):
+            return
 
-        if mask_vars:
-            mask = (
-                f"{next(iter(mask_vars))}"
-                if len(mask_vars) == 1
-                # sorted for deterministic order
-                else f"({' & '.join(sorted(map(str, mask_vars)))})"
-            )
-        return mask
+        assert isinstance(expr, sympy.Expr)
+        indexing = self.indexing(expr, block_ptr=False)
+        assert isinstance(indexing, IndexingOptions)
+
+        index_str = indexing.index_str
+        mask_str = indexing.mask_str if indexing.has_mask() else None
+        size_str = V.kernel.sexpr(self.rename_indexing(size)) if upper else None
+
+        # expr is already wrapped
+        line = self.indirect_assert(
+            index_str, "0" if lower else None, size_str, mask_str
+        )
+
+        indirect = self.is_indirect_indexing(expr) or any(
+            isinstance(m, TritonCSEVariable) for m in indexing.mask_vars
+        )
+        buffer = self.compute if indirect else self.indexing_code
```

Comment:
I think it would be safer to factor out a `get_load_buffer(expr)` so all cases, (e.g. non-persistent reductions) will be handled.

Reply:
I like it better this way.
Non-persistent reductions are also handled in this case, as the `device_assert` will always be before the load. This also handles stores properly and consistently, while it's not clear whether the `get_load_buffer` approach would work there.

Reply:
Load's buffer logic will lift loads out of the `roffset` loop (into `self.body`) when they don't use `rindex`. I would think the assertion would be generated after the load if we don't do the same here?

Reply:
That's alright as that only happens when there is no indirect indexing. In that case, we put the check in `indexing_code`, which comes before `self.body`.

Reply:
`indexing_code` is inside the for loop, whereas `self.body` is outside the loop.

Reply:
Here is an example:
```python
import torch

def fn(a, b):
    idx = torch.arange(a.shape[0], device=a.device) % b.shape[0]
    return (a + b[idx]).sum(-1)

a = torch.randn(1024, 1024, device="cuda")
b = torch.randn(1024, 1, device="cuda")
cfn = torch.compile(fn, dynamic=True)
cfn(a, b)
```

Which generates:
```python
@triton.jit
def triton_(in_ptr0, in_ptr1, out_ptr0, ks0, xnumel, rnumel, XBLOCK : tl.constexpr, RBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:, None]
    xmask = xindex < xnumel
    rbase = tl.arange(0, RBLOCK)[None, :]
    x0 = xindex
    tmp2 = tl.load(in_ptr1 + (x0), xmask, eviction_policy='evict_last')
    _tmp5 = tl.full([XBLOCK, RBLOCK], 0, tl.float32)
    for roffset in range(0, rnumel, RBLOCK):
        rindex = roffset + rbase
        rmask = rindex < rnumel
        r1 = rindex
        tl.device_assert((x0 < ks0) | ~(xmask), "index out of bounds: x0 < ks0")
        tmp0 = tl.load(in_ptr0 + (r1 + (ks0*x0)), rmask & xmask, eviction_policy='evict_first', other=0.0)
        tmp3 = tmp0 + tmp2
        tmp4 = tl.broadcast_to(tmp3, [XBLOCK, RBLOCK])
        tmp6 = _tmp5 + tmp4
        _tmp5 = tl.where(rmask & xmask, tmp6, _tmp5)
    tmp5 = tl.sum(_tmp5, 1)[:, None]
    tl.store(out_ptr0 + (x0), tmp5, xmask)
```

In particular `device_assert((x0 < ks0) ...` should be linked to the load of `b` which is outside the loop.